### PR TITLE
adds a fully safe way how apps can pin files into /tmp temporarily, excepting them from the tmpfiles aging algorithm, based on BSD file locks on dirs we descend into

### PR DESCRIPTION
**Pull request by poettering**
Not sure why we never thought about this before.

This allows us to fix https://github.com/systemd/mkosi/issues/252

### CODE REVIEW

**Comment by poettering** [comment]
An mkosi counterpart PR that makes use of this is here:

https://github.com/systemd/mkosi/pull/319

**Comment by larskarlitski** [comment]
We stumbled across the same issue yesterday (nice timing) in lorax: https://github.com/weldr/lorax/issues/570. Would be cool to see this land soon, so that we can use the same trick. Thanks!

**Comment by poettering** [comment]
let's postpone this to v242, and let's make v241 purely about bugfixing.

File: src/tmpfiles/tmpfiles.c [hunk 1]
```diff
@@ -601,6 +602,11 @@ static int dir_cleanup(
                                         continue;
                                 }
 
+                                if (flock(dirfd(sub_dir), LOCK_EX|LOCK_NB) < 0) {
+                                        log_debug_errno(errno, "Couldn't acquire shared BSD lock on directory \"%s\", skipping: %m", p);
```
Comment (keszybz) [inline src/tmpfiles/tmpfiles.c:607]: I think it'd be better to phrase is less like a failure: "Directory "%s" is locked, skipping."

File: man/tmpfiles.d.xml [hunk 2]
```diff
@@ -545,6 +545,14 @@ w- /proc/sys/vm/swappiness - - - - 10</programlisting></para>
       (ctime). Any of these three (or two) values will prevent cleanup
       if it is more recent than the current time minus the age
       field.</para>
+
+      <para>Note that while the aging algorithm is run a 'shared' BSD file lock (see <citerefentry
```
Comment (keszybz) [inline man/tmpfiles.d.xml:549]: Hmm, "aging algorithm"... Maybe call this "age-based cleanup"?

**Pull request merged**